applications/GeoMechanicsApplication/tests/cpp_tests/test_newmark_quasistatic_U_Pw_scheme.cpp

```diff
@@ -0,0 +1,126 @@
+        result.AddNodalSolutionStepVariable(DT_WATER_PRESSURE);
+        result.AddNodalSolutionStepVariable(DISPLACEMENT);
+        result.AddNodalSolutionStepVariable(WATER_PRESSURE);
+
+        auto p_node = result.CreateNewNode(0, 0.0, 0.0, 0.0);
+        p_node->AddDof(DISPLACEMENT_X);
+        p_node->AddDof(DISPLACEMENT_Y);
+        p_node->AddDof(DISPLACEMENT_Z);
+        p_node->AddDof(WATER_PRESSURE);
+        result.GetProcessInfo()[DELTA_TIME] = 4.0;
+
+        p_node->FastGetSolutionStepValue(VELOCITY, 1) =
+            Kratos::array_1d<double, 3>{1.0, 2.0, 3.0};
+        p_node->FastGetSolutionStepValue(ACCELERATION, 1) =
+            Kratos::array_1d<double, 3>{4.0, 5.0, 6.0};
+        p_node->FastGetSolutionStepValue(DISPLACEMENT, 1) =
+            Kratos::array_1d<double, 3>{7.0, 8.0, 9.0};
+
+        p_node->FastGetSolutionStepValue(WATER_PRESSURE, 1) = 1.0;
+        p_node->FastGetSolutionStepValue(WATER_PRESSURE) = 2.0;
+    }
+
+    ModelPart& GetModelPart()
+    {
+        return mModel.GetModelPart("dummy");
+    }
+};
+
+KRATOS_TEST_CASE_IN_SUITE(CheckScheme_ReturnsZeroForValidScheme, KratosGeoMechanicsFastSuite)
+{
+    NewmarkQuasistaticUPwSchemeTester tester;
+    KRATOS_EXPECT_EQ(tester.mScheme.Check(tester.GetModelPart()), 0);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(InitializeUPWScheme_SetsTimeFactors, KratosGeoMechanicsFastSuite)
+{
+    NewmarkQuasistaticUPwSchemeTester tester;
+    tester.GetModelPart().GetProcessInfo()[DELTA_TIME] = 4.0;
+
+    tester.mScheme.Initialize(tester.GetModelPart());
+
+    // These are the expected numbers according to the SetTimeFactors function
+    constexpr double expected_dt_pressure_coefficient = 1.0 / 3.0;
+    constexpr double expected_velocity_coefficient = 0.5;
+    KRATOS_EXPECT_TRUE(tester.mScheme.SchemeIsInitialized())
+    KRATOS_EXPECT_DOUBLE_EQ(tester.GetModelPart().GetProcessInfo()[DT_PRESSURE_COEFFICIENT],
+                            expected_dt_pressure_coefficient);
+    KRATOS_EXPECT_DOUBLE_EQ(tester.GetModelPart().GetProcessInfo()[VELOCITY_COEFFICIENT],
+                            expected_velocity_coefficient);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(UPWSchemePredict_UpdatesVariablesDerivatives, KratosGeoMechanicsFastSuite)
+{
+    NewmarkQuasistaticUPwSchemeTester tester;
+
+    tester.mScheme.Initialize(tester.GetModelPart()); // This is needed to set the time factors
+    ModelPart::DofsArrayType dof_set;
+    CompressedMatrix A;
+    Vector Dx;
+    Vector b;
+
+    tester.mScheme.Predict(tester.GetModelPart(), dof_set, A, Dx, b);
+
+    // These expected numbers result from the calculations in UpdateVariablesDerivatives
+    const auto expected_acceleration = Kratos::array_1d<double, 3>{-6.75, -9.0, -11.25};
+    const auto expected_velocity = Kratos::array_1d<double, 3>{-4.5, -6.0, -7.5};
+    constexpr auto expected_dt_water_pressure = 1.0 / 3.0;
+
+    KRATOS_EXPECT_EQ(tester.GetModelPart().Nodes()[0].FastGetSolutionStepValue(ACCELERATION),
```

Comment:
I'm not sure whether this macro takes into account a tolerance when comparing the `double` values contained in the arrays. If not, then there is a risk that the test is/becomes fragile (due to comparing (computed) floating-point number for equality).